src/mirror.c

```diff
@@ -0,0 +1,233 @@
+	fprintf(stderr, "   -u, --unset             unset mirror url\n");
+	fprintf(stderr, "   -p, --path=[PATH...]    Use [PATH...] as the path to set/unset the mirror url for (eg: a chroot or btrfs subvol\n");
+	fprintf(stderr, "\n");
+}
+
+static const struct option prog_opts[] = {
+	{ "help", no_argument, 0, 'h' },
+	{ "set", required_argument, 0, 's' },
+	{ "unset", no_argument, 0, 'u' },
+	{ "path", required_argument, 0, 'p' },
+	{ 0, 0, 0, 0 }
+};
+
+static int unset_mirror_url()
+{
+	char *fullpath;
+	if (path_prefix != NULL) {
+		fullpath = mk_full_filename(path_prefix, "/etc/swupd");
+	} else {
+		string_or_die(&fullpath, "/etc/swupd");
+	}
+
+	assert(fullpath != NULL);
+	assert(strcmp(fullpath, "/") != 0);
+	assert(strcmp(fullpath, "/etc") != 0);
+	return swupd_rm(fullpath);
+}
+
+static int write_to_path(char *content, char *path)
+{
+	char *dir, *tmp = NULL;
+	string_or_die(&tmp, "%s", path);
+	dir = dirname(tmp);
+	/* attempt to  make the directory
+	 * ignore EEXIST errors */
+	int ret = mkdir(dir, S_IRWXU);
+	char *cmd;
+	string_or_die(&cmd, "mkdir -p %s", dir);
+	ret = system(cmd);
+	if (ret) {
+		free_string(&cmd);
+		free_string(&tmp);
+		return ret;
+	}
+	free_string(&cmd);
+	free_string(&tmp);
+
+	/* now try to open the file to write */
+	FILE *fp = NULL;
+	fp = fopen(path, "w");
+	if (fp == NULL) {
+		return 1;
+	}
+
+	/* and write to the file */
+	ret = fputs(content, fp);
+	if (ret < 0 || ret == EOF) {
+		fclose(fp);
+		return 1;
+	}
+
+	fclose(fp);
+	return 0;
+}
+
+static int set_mirror_url(char *url)
+{
+	char *content_path;
+	char *version_path;
+	int ret = 0;
+	bool need_unset = false;
+	/* concatenate path_prefix and configuration paths if necessary */
+	if (path_prefix != NULL) {
+		content_path = mk_full_filename(path_prefix, MIRROR_CONTENT_URL_PATH);
+		version_path = mk_full_filename(path_prefix, MIRROR_VERSION_URL_PATH);
+	} else {
+		string_or_die(&content_path, "%s", MIRROR_CONTENT_URL_PATH);
+		string_or_die(&version_path, "%s", MIRROR_VERSION_URL_PATH);
+	}
+
+	/* write url to path_prefix/MIRROR_CONTENT_URL_PATH */
+	ret = write_to_path(url, content_path);
+	if (ret != 0) {
+		need_unset = true;
+		goto out;
+	}
+
+	/* write url to path_prefix/MIRROR_VERSION_URL_PATH */
+	ret = write_to_path(url, version_path);
+	if (ret != 0) {
+		need_unset = true;
+		goto out;
+	}
+
+out:
+	if (need_unset) {
+		unset_mirror_url();
+	}
+	free_string(&content_path);
+	free_string(&version_path);
+	return ret;
+}
+
+char *set = NULL;
+bool unset = false;
+
+static int parse_options(int argc, char **argv)
+{
+	int opt, ret = 0;
+
+	while ((opt = getopt_long(argc, argv, "hs:up:", prog_opts, NULL)) != -1) {
+		switch (opt) {
+		case '?':
+			ret = EINVALID_OPTION;
+			goto out;
+		case 'h':
+			print_help(argv[0]);
+			exit(EXIT_SUCCESS);
+		case 'p':
+			if (!set_path_prefix(optarg)) {
+				fprintf(stderr, "Invalid --path argument\n\n");
+				ret = EINVALID_OPTION;
+				goto out;
+			}
+			break;
+		case 's':
+			if (unset) {
```

Comment:
I think if I was doing this I would have a function pointer to point to either a wrapper around `set_mirror_url` or `unset_mirror_url`, and make the error message more general than `"cannot set and unset at the same time"`, perhaps `"you can only have a maximum of one set ot unset"`. As it is you can `-s` a dozen times without warning.